### DISCUSSION TITLE
How to link additional libraries

### DISCUSSION THREAD

**Discussion by avitase** [friendlyanon/cmake-init#71] [Q&A] [answered]
I have initialized a C project "foobar" and made the following changes:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2e4e720..ea04ae4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -53,6 +53,8 @@ target_include_directories(
     "$<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/export>"
 )
 
+target_link_libraries(foobar_foobar PRIVATE m)
+
 target_compile_features(foobar_foobar PUBLIC c_std_11)
 
 # ---- Install rules ----
diff --git a/source/foobar.c b/source/foobar.c
index 362d5fe..c0bf09d 100644
--- a/source/foobar.c
+++ b/source/foobar.c
@@ -1,6 +1,8 @@
 #include "foobar/foobar.h"
+#define _USE_MATH_DEFINE
+#include <math.h>
 
 const char* exported_function()
 {
-  return "foobar";
+  return M_PI < 3. ? "foobar" : "palim";
 }

```
Obviously, my attempt is to link `libm` to the project but running
```
> cmake --preset=dev-unix
> cmake --build build/dev-unix
```
fails:
```
foobar/source/foobar.c:6:10: error: ‘M_PI’ undeclared (first use in this function)
    6 |   return M_PI < 3. ? "foobar" : "palim";
      |
```
Could you help me to fix this problem?

Thanks!

**Comment by friendlyanon** [comment, marked as answer]
`M_PI` is a non-standard feature of some C standard libraries. On Windows for example, `_USE_MATH_DEFINES` is how you enable this extension: https://docs.microsoft.com/en-us/cpp/c-runtime-library/math-constants?view=msvc-170

On GNU systems the same can be done with `_XOPEN_SOURCE`: https://man7.org/linux/man-pages/man7/feature_test_macros.7.html#:~:text=the%20man%20pages.-,_XOPEN_SOURCE,-Defining%20this%20macro

I recommend not relying to non-standard extensions if not necessary.

Regarding `libm`, it's a Linux-only thing and I'm not totally certain how it could be best handled. Maybe something along the lines of:

```cmake
if(CMAKE_SYSTEM_NAME MATCHES Linux) # probably right?
  include(CheckLibraryExists)
  check_library_exists(m pow "" LIBM_FOUND)

  if(LIBM_FOUND)
    target_link_libraries(project_target PRIVATE m)
  endif()
endif()
```

would be the way to handle it, but I'm not sure.

**Reply by avitase** [reply]
Oh. I just figured out that part of my confusion was triggered by the fact that `gcc` seems to automatically link `libm` (on Linux) but `clang` does not:
```
> rm -rf build/ && CC=`which gcc` cmake --preset=dev-unix && cmake --build build/dev-unix     <-- this links fine
> rm -rf build/ && CC=`which clang` cmake --preset=dev-unix && cmake --build build/dev-unix   <-- this fails!
```
just in case someone finds this helpful :see_no_evil:

**Reply by friendlyanon** [reply]
Yeah, it's probably better to have the user pass in `-lm` via [CMAKE_EXE_LINKER_FLAGS](https://cmake.org/cmake/help/latest/variable/CMAKE_EXE_LINKER_FLAGS.html) (or the MODULE and SHARED variants) like so:

```sh
cmake -B build '-DCMAKE_EXE_LINKER_FLAGS=-lm'
```
